File: pycom/python_matlab_comparison/scripts/cpu/time_sweep_basic_cpu.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov  5 16:14:54 2019

@author: aweiss
"""

from pycom.base.OperationTimer import fancy_timeit_matrix_sweep
import numpy as np
import os

#this allows setting of a global debug
try:
    from __main__ import DEBUG
    if DEBUG is None:
        raise Exception("DEBUG is None")
except:
    DEBUG = False #debug mode for testing
    print("DEBUG Not defined. Setting to False")
if DEBUG:
    print("Debugging Mode Enabled")
    
# what directory to store the data in
try:
    from __main__ import output_directory
except:
    output_directory = './data'


#%% intialize sweep size and repeats
if DEBUG:
    mat_dim_list = np.floor(np.linspace(1,500,6)).astype(np.uint32);
    num_reps = 10
else:
    mat_dim_list = np.floor(np.linspace(1,10000,51)).astype(np.uint32);
    num_reps = 100

out_file_np = os.path.join(output_directory,'stats_py_np_basic.mat')
out_file_nb = os.path.join(output_directory,'stats_py_nb_basic.mat')

#%%
'''
NUMPY TESTS
'''
#%% Test the fancy_timeit_matrix_sweep capability
combo_fun   = lambda x,y: x+y*np.exp(x)
functs      = [np.add,np.subtract,np.multiply,np.divide,np.exp,combo_fun]
funct_names = ['add' ,'sub'      ,'mult'     ,'div'    ,'exp' ,'combo'  ]
num_args    = [2     ,2          ,2          ,2        ,1     ,2        ]

#% now run 
[stats_np_double,_] = fancy_timeit_matrix_sweep(
	functs,funct_names,num_args,mat_dim_list,num_reps,dtype=np.cdouble);
[stats_np_single,_] = fancy_timeit_matrix_sweep(
	functs,funct_names,num_args,mat_dim_list,num_reps,dtype=np.csingle);

#%%
'''
NUMBA TESTS
'''
#%% Function declarations
import cmath
from numba import vectorize,complex128,complex64

@vectorize([complex64(complex64,complex64),complex128(complex128,complex128)],target='parallel')
def nb_add(x,y):
    return x+y

@vectorize([complex64(complex64,complex64),complex128(complex128,complex128)],target='parallel')
def nb_subtract(x,y):
    return x+y

@vectorize([complex64(complex64,complex64),complex128(complex128,complex128)],target='parallel')
def nb_multiply(x,y):
    return x+y

@vectorize([complex64(complex64,complex64),complex128(complex128,complex128)],target='parallel')
def nb_divide(x,y):
    return x+y

@vectorize([complex64(complex64),complex128(complex128)],target='parallel')
def nb_exp(x):
    return cmath.exp(x)

@vectorize([complex64(complex64,complex64),complex128(complex128,complex128)],target='parallel')
def nb_comb(x,y):
    return x+y*cmath.exp(x)

#%% Test the fancy_timeit_matrix_sweep capability
#%add/sub/mult/div
functs       = [nb_add,nb_subtract,nb_multiply,nb_divide,nb_exp,nb_comb];
funct_names  = ['add' ,'sub'      ,'mult'     ,'div'    ,'exp' ,'combo'];
num_args     = [2     ,2          ,2          ,2        ,1     ,2      ];

#% now run
[stats_nb_double,_] = fancy_timeit_matrix_sweep(
	functs,funct_names,num_args,mat_dim_list,num_reps,dtype=np.cdouble); 
[stats_nb_single,_] = fancy_timeit_matrix_sweep(
	functs,funct_names,num_args,mat_dim_list,num_reps,dtype=np.csingle); 
    
#%% now save out
import scipy.io as sio
sio.savemat(out_file_np,{'double':stats_np_double,'single':stats_np_single})
sio.savemat(out_file_nb,{'double':stats_nb_double,'single':stats_nb_single})

    

File: pycom/python_matlab_comparison/scripts/cpu/time_sweep_basic_cpu.m
%% some setup
if ~exist('pycom_path','var') % check for pycom path
    file_dir = fileparts(mfilename('fullpath')); %get the directory of this file
    pycom_path = fullfile(file_dir,'../../'); %assume the base is 2 levels up from this
end

if ~exist('DEBUG','var') % Check for debug flag
    DEBUG = false;
    fprintf("DEBUG Not defined. Setting to False");
end

if ~exist('output_directory','var') % check for provided output directory
    output_directory = './data';
end

addpath(fullfile(pycom_path,'base')); % add OperationTimer

%% now lets setup and run our test

if DEBUG
    dim_list = floor(linspace(1,500,6));
    num_reps = 10;
else
    dim_list = floor(linspace(1,10000,51));
    num_reps = 100;
end
out_file = fullfile(output_directory,'stats_mat_basic.mat');

%% now run
mat_dim_list = dim_list;

combo_fun   = @(x,y) x+y.*exp(x);
functs      = {@plus ,@minus,@times ,@rdivide,@exp  ,combo_fun};
funct_names = {'add' ,'sub' ,'mult' ,'div'   ,'exp' ,'combo'  };
num_args    = [2    ,2    ,2     ,2      ,1    ,2        ];

% now run 
stats_double = OperationTimer.fancy_timeit_matrix_sweep(...
	functs,funct_names,num_args,mat_dim_list,num_reps,'dtype',@double);
stats_single = OperationTimer.fancy_timeit_matrix_sweep(...
	functs,funct_names,num_args,mat_dim_list,num_reps,'dtype',@single);

%matlab cant have two things named the same...
m_single = stats_single;
m_double = stats_double;

save(out_file,'m_single','m_double');



    